obkvrpc/compressor.go

```diff
@@ -0,0 +1,155 @@
+ * EITHER EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO NON-INFRINGEMENT,
+ * MERCHANTABILITY OR FIT FOR A PARTICULAR PURPOSE.
+ * See the Mulan PSL v2 for more details.
+ * #L%
+ */
+
+package obkvrpc
+
+import (
+	"bytes"
+	"compress/zlib"
+	"fmt"
+	"github.com/golang/snappy"
+	"github.com/klauspost/compress/zstd"
+	"github.com/oceanbase/obkv-table-client-go/protocol"
+	"github.com/pierrec/lz4"
+	"github.com/pkg/errors"
+	"io"
+)
+
+type DeCompressor interface {
+	Decompress(src *bytes.Buffer, originLen int32) (*bytes.Buffer, error)
+}
+
+type NoneDecompressor struct{}
+type Lz4Decompressor struct{}
+type SnappyDecompressor struct{}
+type ZlibDecompressor struct{}
+type ZStdDecompressor struct{}
+
+func (d *NoneDecompressor) Decompress(src *bytes.Buffer, originLen int32) (*bytes.Buffer, error) {
+	return src, nil
+}
+
+func (d *Lz4Decompressor) Decompress(src *bytes.Buffer, originLen int32) (*bytes.Buffer, error) {
+	srcBytes := src.Bytes()
+	dstBytes := make([]byte, originLen)
+	n, err := lz4.UncompressBlock(srcBytes, dstBytes)
+	if err != nil {
+		return nil, err
+	}
+	if int32(n) != originLen {
+		return nil, errors.New(fmt.Sprintf("length of decompress content is not expected: actual-%d, expected-%d", n, originLen))
+	}
+	return bytes.NewBuffer(dstBytes), nil
+}
+
+func (d *SnappyDecompressor) Decompress(src *bytes.Buffer, originLen int32) (*bytes.Buffer, error) {
+	srcBytes := src.Bytes()
+	dstBytes, err := snappy.Decode(nil, srcBytes)
+	if err != nil {
+		return nil, err
+	}
+	if int32(len(dstBytes)) != originLen {
+		return nil, errors.New(fmt.Sprintf("length of decompress content is not expected: actual-%d, expected-%d", int32(len(dstBytes)), originLen))
+	}
+	return bytes.NewBuffer(dstBytes), nil
+}
+
+func (d *ZlibDecompressor) Decompress(src *bytes.Buffer, originLen int32) (*bytes.Buffer, error) {
+	// 1. create zlib reader
+	zlibReader, err := zlib.NewReader(src)
+	if err != nil {
+		return nil, err
+	}
+
+	defer func(zlibReader io.ReadCloser) {
+		err := zlibReader.Close()
+		if err != nil {
+			panic(err.Error())
+		}
+	}(zlibReader)
+
+	// 2. create buffer
+	dstBuffer := new(bytes.Buffer)
+	// 3. write the decompress data to dstBuffer
+	n, err := io.Copy(dstBuffer, zlibReader)
+	if err != nil {
+		return nil, err
+	}
+	if int32(n) != originLen {
+		return nil, errors.New(fmt.Sprintf("length of decompress content is not expected: actual-%d, expected-%d", n, originLen))
+	}
+	return dstBuffer, nil
+}
+
+func (d *ZStdDecompressor) Decompress(src *bytes.Buffer, originLen int32) (*bytes.Buffer, error) {
+	zstdReader, err := zstd.NewReader(src)
+	if err != nil {
+		return nil, err
+	}
+	defer zstdReader.Close()
+	dstBuffer := new(bytes.Buffer)
+	n, err := io.Copy(dstBuffer, zstdReader)
+	if err != nil {
+		return nil, err
+	}
+	if int32(n) != originLen {
+		return nil, errors.New(fmt.Sprintf("length of decompress content is not expected: actual-%d, expected-%d", n, originLen))
```

Comment:
ZStdDecompressor: length of decompress content ...

Reply:
done